mergify_engine/worker.py

```diff
@@ -0,0 +1,366 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+import asyncio
+import collections
+import concurrent.futures
+import contextlib
+import datetime
+import functools
+import itertools
+import signal
+import time
+
+import daiquiri
+import msgpack
+import uvloop
+
+from mergify_engine import config
+from mergify_engine import engine
+from mergify_engine import exceptions
+from mergify_engine import utils
+from mergify_engine.clients import github
+
+
+LOG = daiquiri.getLogger(__name__)
+
+
+MAX_RETRIES = 3
+
+WORKER_PROCESSING_DELAY = 30
+
+
+class PullRetry(Exception):
+    def __init__(self, attempts):
+        self.attempts = attempts
+        super().__init__(self)
+
+
+class MaxPullRetry(PullRetry):
+    pass
+
+
+class StreamRetry(Exception):
+    def __init__(self, attempts, retry_at):
+        self.attempts = attempts
+        self.retry_at = retry_at
+        super().__init__(self)
+
+
+class MaxStreamRetry(StreamRetry):
+    pass
+
+
+def push(installation_id, owner, repo, pull_number, event_type, data):
+    redis = utils.get_redis_for_stream()
+    stream_name = f"stream~{installation_id}".encode()
+    scheduled_at = utils.utcnow() + datetime.timedelta(seconds=WORKER_PROCESSING_DELAY)
```

Comment:
does that mean you need 30s *without* event to see things processed?
might be a bit long in some cases

Reply:
Because of `nx=True`, the date is used only if the stream is empty.
So it's 30 seconds after the first event of the streams.